File: anchor-reference-server/src/main/java/org/stellar/anchor/reference/event/ActiveTransactionStore.java
package org.stellar.anchor.reference.event;

import java.util.Set;

/** Stores the transactions that are currently being processed by the anchor. */
public interface ActiveTransactionStore {

  /**
   * Adds a transaction to the store.
   *
   * @param customerId The customer associated with the transaction.
   * @param transactionId The transaction id.
   * @return The set of transactions that are currently being processed for the account.
   */
  Set<String> add(String customerId, String transactionId);

  /**
   * Removes a transaction from the store.
   *
   * @param customerId The customer associated with the transaction.
   * @param transactionId The transaction id.
   */
  void remove(String customerId, String transactionId);

  /**
   * Gets the transactions that are currently being processed for the account.
   *
   * @param customerId The customer associated with the transaction.
   * @return The set of transactions that are currently being processed for the account.
   */
  Set<String> getTransactions(String customerId);
}

File: kotlin-reference-server/src/main/kotlin/org/stellar/reference/event/processor/ActiveTransactionStore.kt
package org.stellar.reference.event.processor

interface ActiveTransactionStore {
  fun addTransaction(customerId: String, transactionId: String)
  fun removeTransaction(customerId: String, transactionId: String)
  fun getTransactions(customerId: String): List<String>
}

class InMemoryTransactionStore : ActiveTransactionStore {
  private val transactions = mutableMapOf<String, MutableSet<String>>()

  override fun addTransaction(customerId: String, transactionId: String) {
    transactions.getOrPut(customerId) { mutableSetOf() }.add(transactionId)
  }

  override fun removeTransaction(customerId: String, transactionId: String) {
    transactions[customerId]?.remove(transactionId)
  }

  override fun getTransactions(customerId: String): List<String> {
    return transactions[customerId]?.toList() ?: emptyList()
  }
}
